### ISSUE TITLE
Windows Latest on Python 3.8 Constantly Fails

### ISSUE THREAD

**Issue by adam-grant-hendry** [snok/install-poetry#94]
For some reason, Windows-2022 is now failing for me on Python 3.8 only:

```
Run snok/install-poetry@v[1](https://github.com/adam-grant-hendry/qtpygraph/actions/runs/3202205327/jobs/5230946943#step:6:1).3.2
Run $GITHUB_ACTION_PATH/main.sh

Setting Poetry installation path as C:/Users/runneradmin/AppData/Roaming/Python/Scripts

Installing Poetry 👷

Retrieving Poetry metadata

# Welcome to Poetry!

This will download and install the latest version of Poetry,
a dependency and package manager for Python.

It will add the `poetry` command to Poetry's bin directory, located at:

C:\Users\runneradmin\AppData\Roaming\Python\Scripts\bin

You can uninstall at any time by executing this script with the --uninstall option,
and these changes will be reverted.

Installing Poetry (1.2.1)
Installing Poetry (1.2.1): Creating environment
Installing Poetry (1.2.1): An error occurred. Removing partial environment.
Traceback (most recent call last):
  File "C:/Users/RUNNER~1/AppData/Local/Temp/tmp.0IGyZ1qQC7", line 9[40](https://github.com/adam-grant-hendry/qtpygraph/actions/runs/3202205327/jobs/5230946943#step:6:42), in <module>
    sys.exit(main())
  File "C:/Users/RUNNER~1/AppData/Local/Temp/tmp.0IGyZ1qQC7", line 919, in main
    return installer.run()
  File "C:/Users/RUNNER~1/AppData/Local/Temp/tmp.0IGyZ1qQC7", line 550, in run
    self.install(version)
  File "C:/Users/RUNNER~1/AppData/Local/Temp/tmp.0IGyZ1qQC7", line 571, in install
    with self.make_env(version) as env:
  File "C:\hostedtoolcache\windows\Python\3.8.10\x64\lib\contextlib.py", line 113, in __enter__
    return next(self.gen)
  File "C:/Users/RUNNER~1/AppData/Local/Temp/tmp.0IGyZ1qQC7", line 643, in make_env
    raise e
  File "C:/Users/RUNNER~1/AppData/Local/Temp/tmp.0IGyZ1qQC7", line 629, in make_env
    yield VirtualEnvironment.make(env_path)
  File "C:/Users/RUNNER~1/AppData/Local/Temp/tmp.0IGyZ1qQC7", line 319, in make
    builder.create(target)
  File "C:\hostedtoolcache\windows\Python\3.8.10\x64\lib\venv\__init__.py", line 68, in create
    self._setup_pip(context)
  File "C:\hostedtoolcache\windows\Python\3.8.10\x64\lib\venv\__init__.py", line 289, in _setup_pip
    subprocess.check_output(cmd, stderr=subprocess.STDOUT)
  File "C:\hostedtoolcache\windows\Python\3.8.10\x64\lib\subprocess.py", line [41](https://github.com/adam-grant-hendry/qtpygraph/actions/runs/3202205327/jobs/5230946943#step:6:43)5, in check_output
    return run(*popenargs, stdout=PIPE, timeout=timeout, check=True,
  File "C:\hostedtoolcache\windows\Python\3.8.10\x64\lib\subprocess.py", line [49](https://github.com/adam-grant-hendry/qtpygraph/actions/runs/3202205327/jobs/5230946943#step:6:51)3, in run
    with Popen(*popenargs, **kwargs) as process:
  File "C:\hostedtoolcache\windows\Python\3.8.10\x64\lib\subprocess.py", line 8[58](https://github.com/adam-grant-hendry/qtpygraph/actions/runs/3202205327/jobs/5230946943#step:6:60), in __init__
    self._execute_child(args, executable, preexec_fn, close_fds,
  File "C:\hostedtoolcache\windows\Python\3.8.10\x[64](https://github.com/adam-grant-hendry/qtpygraph/actions/runs/3202205327/jobs/5230946943#step:6:66)\lib\subprocess.py", line 1311, in _execute_child
    hp, ht, pid, tid = _winapi.CreateProcess(executable, args,
FileNotFoundError: [WinError 2] The system cannot find the file specified
Error: Process completed with exit code 1.
```

**Comment by sondrelg** [comment]
Thanks for the report. Does changing v1 to v1.3.1 (the previous release) fix the issue for you?

**Comment by adam-grant-hendry** [comment]
@sondrelg How strange. Yes! That completely fixed it!

...What's happened from v1.3.1 to v1.3.2?

**Comment by adam-grant-hendry** [comment]
Specifically, the error _only_ happened for Windows and version 3.8. MacOS and Linux on 3.8 worked and 3.9 and 3.10 on all OS's worked.

**Comment by sondrelg** [comment]
We mostly just quoted env vars in the latest release. I'll add a test case for Windows later to see if I can reproduce. 

Would you be able to share the relevant parts of your workflow to make that easier?

**Comment by adam-grant-hendry** [comment]
@sondrelg It's a public repo, which makes it easier. Here's the link: https://github.com/adam-grant-hendry/qtpygraph

**Comment by adam-grant-hendry** [comment]
@sondrelg Feel free to look at my workflows

**Comment by miigotu** [comment]
@sondrelg 
Probably should add python 3.8 back to the test matrix?
https://github.com/snok/install-poetry/compare/v1.3.1..v1.3.2?diff=unified#diff-faff1af3d8ff408964a57b2e475f69a6b7c7b71c9978cccc8f471798caac2c88R12

Also, there is an open bug about this script not working with windows:
https://github.com/snok/install-poetry/compare/v1.3.1..v1.3.2?diff=unified#diff-f121df8f1f148ef72415f98dd27089a0e6cec3ec939840286358170e5e20c990R6

Ill try to find it. I think if you change this back
```diff
++curl -sSL https://raw.githubusercontent.com/python-poetry/poetry/48339106eb0d403a3c66519317488c8185844b32/install-poetry.py --output "$installation_script"
--curl -sSL https://install.python-poetry.org/ --output "$installation_script"
```
The problem will be fixed/

**Comment by miigotu** [comment]
Removing -sSL and adding `--ssl-no-revoke ` to the installer arguments supposedly fixes it.
https://github.com/python-poetry/poetry/issues/6117

**Comment by miigotu** [comment]
Here it is: https://github.com/python-poetry/install.python-poetry.org/issues/46

**Comment by adam-grant-hendry** [comment]
> If you remove the shell default it will work

Interesting...that's been part of what was recommended previously for this action (i.e. to use `bash` and the default shell). @sondrelg isn't that true? Does this need to change going forward?

**Comment by sondrelg** [comment]
Of course there's a regression specifically for 3.8 just as we remove it from the matrix 😅 It's a bit of a pain to have 20+ permutations of the basic install test, but I suppose this is a prime example of why it's worth having.

Yes @adam-grant-hendry, IIRC, removing the `bash` shell default on windows causes other problems. Not sure if they've gone away yet, but would be good to know.

Either way, I suppose we could just switch to `--ssl-no-revoke` if that fixes it for all platforms. Don't think certificate revocation checks are relevant when curling the hardcoded official installation script, like we're doing here. Do you agree?

**Comment by sondrelg** [comment]
Surprisingly, CI doesn't fail when including windows 3.8 ([here](https://github.com/snok/install-poetry/actions/runs/3202445864/jobs/5231437435)) 🤔

**Comment by miigotu** [comment]
> Surprisingly, CI doesn't fail when including windows 3.8 ([here](https://github.com/snok/install-poetry/actions/runs/3202445864/jobs/5231437435)) thinking

If you add this to the test.yml it will fail
https://github.com/SickChill/sickchill/blob/test-deploy/.github/workflows/pythonpackage.yml#L27L29

**Comment by sondrelg** [comment]
There is this already

https://github.com/snok/install-poetry/blob/main/.github/workflows/test.yml#L58:L60

Are you saying the location of the default specification matters?

**Comment by adam-grant-hendry** [comment]
> Either way, I suppose we could just switch to --ssl-no-revoke if that fixes it for all platforms. Don't think certificate revocation checks are relevant when curling the hardcoded official installation script, like we're doing here. Do you agree?

I'm perfectly fine sticking with `v1.3.1` right now. I don't have an immediate need to upgrade. I personally used `default: bash` only because it was easier to stick to one scripting language instead of having to switch back-and-forth between that and powershell.

Not sure I can say I know much about the `--ssl-no-revoke` option or how it works, but if it works I'm fine with it! 😆

**Comment by adam-grant-hendry** [comment]
> Are you saying the location of the default specification matters?

Looks like yours is specific to a job, whereas mine and @miigotu have ours for the entire workflow. Is that correct?

**Comment by miigotu** [comment]
> There is this already
> 
> https://github.com/snok/install-poetry/blob/main/.github/workflows/test.yml#L58:L60
> 
> Are you saying the location of the default specification matters?

I'm guessing because that job isnt creating a venv (https://github.com/snok/install-poetry/blob/main/.github/workflows/test.yml#L117) it skips the bug

**Comment by miigotu** [comment]
In that issue they said its only with msys bash on windows, I have to check OP logs but here it is git bash
https://github.com/snok/install-poetry/actions/runs/3202507531/jobs/5231559465#step:3:104

**Comment by miigotu** [comment]
I wonder if it is him setting: `SHELLOPTS: errexit:pipefail` and we do that also, causing some weirdness? I dont see anything else really different with the environment.

**Comment by sondrelg** [comment]
So far I've tried:

- Setting Python to 3.8.10 which is the version specifically used in the failing run above
- Added `SHELLOPTS` to env
- Tried setting windows-2022 instead of windows-latest

but so far nothing reproduces

**Comment by sondrelg** [comment]
In the latest push to #95 I've added the `--ssl-no-revoke` command to main.sh.

@adam-grant-hendry, would you mind testing with `@94` instead of `v1.3.1` to see if that works for you?

**Comment by GuillaumeFalourd** [comment]
I was having with the same issue than @adam-grant-hendry in my workflow using:

```
      - name: Install and configure Poetry
        uses: snok/install-poetry@v1
        with:
          version: 1.1.10
          virtualenvs-create: false
```

and (after trying to update to the latest version) with:

```
      - name: Install and configure Poetry
        uses: snok/install-poetry@v1.3.2
        with:
          version: 1.2.1
          virtualenvs-create: false
```

After reading this thread and seeing the tests workflows @sondrelg shared, I used instead:

```
      - name: Install and configure Poetry
        uses: snok/install-poetry@v1.3.1
        with:
          version: 1.1.14
          virtualenvs-create: false
```

And the error didn't occur anymore 🤷‍♂️

**Comment by miigotu** [comment]
We need to find the cause though. Can you try updating just one or the other of those? Aka is the problem here, or in poetry itself?

**Comment by GuillaumeFalourd** [comment]
@miigotu: It worked on the windows-latest runner with the action version `1.3.1`, but **not** with the version `1.3.2` using poetry version `1.1.14`

**Comment by sondrelg** [comment]
Then perhaps the changes made to https://github.com/snok/install-poetry/blob/main/main.sh#L38:L40 is the reason. 

The thing we need to get to the bottom of this is a reproducible example of a failing workflow for v1.3.2. So far we haven't been able to reproduce the issue. Any ideas as to why not @GuillaumeFalourd?

**Comment by GuillaumeFalourd** [comment]
I've got a workflow reproducing the error [here](https://github.com/GuillaumeFalourd/poc-github-actions/actions/runs/3219699229) using this [workflow implementation](https://github.com/GuillaumeFalourd/poc-github-actions/blob/main/.github/workflows/workflow-tester58.yml).

**Comment by GuillaumeFalourd** [comment]
After making some tests in a fork, it seems the problem is coming from this commit [https://github.com/snok/install-poetry/commit/1218f0b98ae913ee2f2ad64868af8a3ca66511b9], by adding the `set -eo pipefail` line.

Just removing this line from the `main.sh` made the previous workflow which was failing working again: [evidence](https://github.com/GuillaumeFalourd/install-poetry/actions/runs/3221930536/jobs/5270440194)

**Comment by sondrelg** [comment]
Great, thanks for that! I'll do a little bit more testing, then see if we can release a fix in a few hours 👍

**Comment by miigotu** [comment]
https://askubuntu.com/questions/1044280/set-eo-pipefail-not-working-in-windows-subsystem-for-linux-ubuntu-16-04

**Comment by sondrelg** [comment]
Having had time to look at https://github.com/python-poetry/install.python-poetry.org/issues/46 @miigotu, I think they're saying git bash on windows runners are affected, so I think this is an upstream bug. The script target was changed a while back, but I hadn't released a new version until v1.3.2, so that's why this is happening.

Perhaps we should do a simple
```
if windows on python 3.8 or less:
    install using old script
else:
    install using new script
```

What do you think?
